File: web/src/catalog/enchants.ts
import { ItemSlot } from '../gen/wowfsim/sim_pb.ts'
import type { CatalogItem } from './era.ts'
import raw from './enchants.json' with { type: 'json' }

export type CatalogEnchant = {
  id: number
  name: string
  icon: string
  quality?: number
  kind: 'permanent' | 'temporary' | string
  slots: number[]
  effectName: string
  effectLabel: string
  useText: string
  twoHandOnly?: boolean
  shieldOnly?: boolean
  offHandOnly?: boolean
  weapon?: boolean
  ranged?: boolean
  proc?: boolean
  strength?: number
  agility?: number
  stamina?: number
  intellect?: number
  spirit?: number
  attackPower?: number
  spellPower?: number
  healingPower?: number
  critChance?: number
  hitChance?: number
  spellCritChance?: number
  spellHitChance?: number
  haste?: number
  weaponDamage?: number
  mp5?: number
}

export const ENCHANTS: CatalogEnchant[] = raw as CatalogEnchant[]

export function enchantById(id: number): CatalogEnchant | undefined {
  return ENCHANTS.find((entry) => entry.id === id)
}

export function enchantFitsSlot(enchant: CatalogEnchant, slot: ItemSlot, item?: CatalogItem | null): boolean {
  if (!enchant.slots.includes(slot)) {
    return false
  }
  if (enchant.twoHandOnly && item && item.hand !== '2h') {
    return false
  }
  if (enchant.shieldOnly && item && item.itemSubclass !== 'Shield') {
    return false
  }
  if (enchant.offHandOnly && item && (item.hand !== 'oh' || item.itemSubclass === 'Shield')) {
    return false
  }
  if (enchant.weapon && slot === ItemSlot.OFF_HAND && item?.itemSubclass === 'Shield') {
    return false
  }
  if (enchant.ranged && slot !== ItemSlot.RANGED) {
    return false
  }
  return true
}

export function enchantsForSlot(
  slot: ItemSlot,
  item?: CatalogItem | null,
  kind?: string,
): CatalogEnchant[] {
  return ENCHANTS.filter((enchant) => {
    if (kind && enchant.kind !== kind) {
      return false
    }
    return enchantFitsSlot(enchant, slot, item)
  })
}

export function slotHasPermanentEnchants(slot: ItemSlot) {
  return ENCHANTS.some((enchant) => enchant.kind === 'permanent' && enchant.slots.includes(slot))
}

File: engine/internal/clientdata/enchants.go
package clientdata

import (
	_ "embed"
	"encoding/json"
	"sync"
)

//go:embed enchants.json
var enchantsJSON []byte

type Enchant struct {
	ID               int32   `json:"id"`
	Name             string  `json:"name"`
	Icon             string  `json:"icon"`
	Quality          int32   `json:"quality"`
	Kind             string  `json:"kind"`
	Slots            []int32 `json:"slots"`
	EffectName       string  `json:"effectName"`
	EffectLabel      string  `json:"effectLabel"`
	UseText          string  `json:"useText"`
	TwoHandOnly      bool    `json:"twoHandOnly"`
	ShieldOnly       bool    `json:"shieldOnly"`
	OffHandOnly      bool    `json:"offHandOnly"`
	Weapon           bool    `json:"weapon"`
	Ranged           bool    `json:"ranged"`
	Proc             bool    `json:"proc"`
	Strength         int32   `json:"strength"`
	Agility          int32   `json:"agility"`
	Stamina          int32   `json:"stamina"`
	Intellect        int32   `json:"intellect"`
	Spirit           int32   `json:"spirit"`
	AttackPower      int32   `json:"attackPower"`
	SpellPower       int32   `json:"spellPower"`
	HealingPower     int32   `json:"healingPower"`
	CritChance       float64 `json:"critChance"`
	HitChance        float64 `json:"hitChance"`
	SpellCritChance  float64 `json:"spellCritChance"`
	SpellHitChance   float64 `json:"spellHitChance"`
	Haste            float64 `json:"haste"`
	WeaponDamage     float64 `json:"weaponDamage"`
	Mp5              int32   `json:"mp5"`
}

var (
	enchantsOnce sync.Once
	enchants     []Enchant
	enchantByID  map[int32]Enchant
)

func loadEnchants() {
	enchantsOnce.Do(func() {
		if err := json.Unmarshal(enchantsJSON, &enchants); err != nil {
			panic(err)
		}
		enchantByID = make(map[int32]Enchant, len(enchants))
		for _, enc := range enchants {
			enchantByID[enc.ID] = enc
		}
	})
}

func EnchantByID(id int32) (Enchant, bool) {
	loadEnchants()
	enc, ok := enchantByID[id]
	return enc, ok
}
